images/virtualization-artifact/pkg/controller/vm/internal/block_device.go

```diff
@@ -237,60 +241,71 @@ func (h *BlockDeviceHandler) checkBlockDevicesSanity(vm *virtv2.VirtualMachine)
 }
 
 // countReadyBlockDevices check if all attached images and disks are ready to use by the VM.
-func (h *BlockDeviceHandler) countReadyBlockDevices(vm *virtv2.VirtualMachine, s BlockDevicesState, log *slog.Logger) (int, bool) {
+func (h *BlockDeviceHandler) countReadyBlockDevices(vm *virtv2.VirtualMachine, s BlockDevicesState, log *slog.Logger) (int, bool, []string) {
 	if vm == nil {
-		return 0, false
+		return 0, false, nil
 	}
+
+	var warnings []string
 	ready := 0
-	viProvisioned := true
-	cviProvisioned := true
-	vdProvisioned := true
+	canStartKVVM := true
 	for _, bd := range vm.Spec.BlockDeviceRefs {
 		switch bd.Kind {
 		case virtv2.ImageDevice:
 			if vi, hasKey := s.VIByName[bd.Name]; hasKey && vi.Status.Phase == virtv2.ImageReady {
 				ready++
 				continue
 			}
-			viProvisioned = false
+			canStartKVVM = false
 		case virtv2.ClusterImageDevice:
 			if cvi, hasKey := s.CVIByName[bd.Name]; hasKey && cvi.Status.Phase == virtv2.ImageReady {
 				ready++
 				continue
 			}
-			cviProvisioned = false
+			canStartKVVM = false
 		case virtv2.DiskDevice:
-			if vd, hasKey := s.VDByName[bd.Name]; hasKey {
-				// we can attach vd to only one vm
-				attachReady := len(vd.Status.AttachedToVirtualMachines) == 0
-				for _, attached := range vd.Status.AttachedToVirtualMachines {
-					if attached.Name != vm.GetName() {
-						attachReady = false
-						msg := fmt.Sprintf("VirtualDisk %s was attached to another virtual machine %q", vd.Name, attached.Name)
-						log.Warn(msg)
-						h.recorder.Event(vm, corev1.EventTypeWarning, virtv2.ReasonVDAlreadyInUse, msg)
-						break
-					}
-					attachReady = true
-				}
+			vd, hasKey := s.VDByName[bd.Name]
+			if !hasKey {
+				canStartKVVM = false
+				continue
+			}
 
-				if !attachReady || vd.Status.Target.PersistentVolumeClaim == "" {
-					continue
+			var canAttach bool
+
+			switch {
```

Comment:
```
				canAttach := true
				for _, attached := range vd.Status.AttachedToVirtualMachines {
					if attached.Name != vm.GetName() {
						canAttach = false
						msg := fmt.Sprintf("VirtualDisk %s was attached to another virtual machine %q", vd.Name, attached.Name)
						log.Warn(msg)
						h.recorder.Event(vm, corev1.EventTypeWarning, virtv2.ReasonVDAlreadyInUse, msg)
					}
				}
```
maybe something like this? 

we check everything and at the same time create an event for each connected disk